File: plot/plot.py
#!/usr/bin/env python

from scipy.special import expi
from numpy import exp, log, linspace, euler_gamma, insert, zeros_like, arange, ceil, where, append
from matplotlib import pyplot as plt

plt.rcParams.update({
	'text.usetex': True,
	'font.size': 10,
	'font.family': 'lmodern',
	'text.latex.preamble': r'\usepackage{lmodern}'
})

def li(x):
	return expi(log(x))

def repow(x, k):
	return where(x>0, x**k, 0.0)

p = linspace(0.001, 0.999, 999)
mu1 = (li(p) - log(-log(p)) - euler_gamma) / log(p)
p = [0, *p, 1]
mu1 = [0, *mu1, 1]

plt.figure(figsize=(6,2))
plt.plot(p, mu1)
plt.ylim(0, 1)
plt.xlabel('$P$')
plt.ylabel(r'$\mu_1(P)$')
plt.savefig('mu1.pdf', bbox_inches='tight', transparent=True, format='pdf')

def x_and_F(P, upper_x = 1.3):
	lnP = log(P)
	step = 0.005
	hard_zero = 0.02 # F(P,x<hard_zero) is hardcoded to zero
	hard_zero_i = int(hard_zero / step)
	x = arange(0, 1, step)
	if upper_x > 1:
		x = append(x, [upper_x])
	F = zeros_like(x)
	s = 1
	s_fac = 1
	while True:
		x1 = arange(hard_zero, 1/s, step)
		if len(x1) == 0:
			break
		s_fac *= s
		F[hard_zero_i:hard_zero_i+len(x1)] += P**(s*x1) * ((1-s*x1)*lnP)**s / s_fac / s
		s += 1
	F -= li(P) - log(-lnP) - euler_gamma
	F /= lnP
	F += x
	F[:hard_zero_i] = 0
	return x, F

def plot1(p):
	plt.plot(*x_and_F(p), label='$P=%.2f$'%p)

plt.figure(figsize=(4.8,3.6))
plot1(0.01)
plot1(0.03)
plot1(0.09)
plot1(0.3)
plot1(0.6)
plot1(0.9)
xmin, xmax = plt.xlim()
plt.xlim(xmin, 1.2)
plt.ylim(0, 0.6)
plt.legend()
plt.xlabel('$x$')
plt.ylabel('$F(P,x)$')
plt.savefig('F.pdf', bbox_inches='tight', transparent=True, format='pdf')

def plot_f_k(k):
	if k == 0:
		plt.plot([0, 1, 1], [0, 0, 5], label='$k=0$')
		return
	if k == 1:
		plt.plot([0, 0.5, 0.5, 1], [0, 0, 2, 2], label='$k=1$')
		return
	x = linspace(0, 1, 201)
	f = 0
	fac = k*(k+1)
	for s in range(1, k+2):
		f += fac * repow(1-s*x, k-1)
		fac *= (k+1-s) / -s
	plt.plot(x, f, label=f'$k={k}$')

plt.figure(figsize=(2.8,2.1))
plot_f_k(0)
plot_f_k(1)
plot_f_k(2)
plot_f_k(3)
plot_f_k(4)
plt.ylim(0, 4)
plt.xlabel('$x$')
plt.ylabel('$f_k(x)$')
plt.savefig('f_k.pdf', bbox_inches='tight', transparent=True, format='pdf')

def x_and_F_k(k, upper_x = 1.3):
	x = linspace(0, 1, 201)
	if upper_x > 1:
		x = append(x, [upper_x])
	F = zeros_like(x)
	fac = 1/(k+1)
	for s in range(1, k+2):
		fac *= (k+2-s) / -s
		F += fac/s * (1 - repow(1-s*x, k+1))
	F += x
	return x, F

def plot_F_k(k):
	plt.plot(*x_and_F_k(k), label=f'$k={k}$')

plt.figure(figsize=(2.8,2.1))
plot_F_k(0)
plot_F_k(1)
plot_F_k(2)
plot_F_k(3)
plot_F_k(4)
xmin, xmax = plt.xlim()
plt.xlim(xmin, 1.2)
plt.ylim(0, 0.6)
plt.legend()
plt.xlabel('$x$')
plt.ylabel('$F_k(x)$')
plt.savefig('BigF_k.pdf', bbox_inches='tight', transparent=True, format='pdf')

def plot2(P):
	lmd = -log(P)
	x, F = x_and_F(P, upper_x = (2 + log(lmd))/lmd)
	y = x * lmd - log(lmd)
	plt.plot(y, F * lmd, label=r'$\mathrm e^{-\lambda}=%.4f$' % P)

plt.figure(figsize=(2.8,2.1))
plot2(0.0001)
plot2(0.001)
plot2(0.01)
plot2(0.1)
y = linspace(-2, 2, 201)
plt.plot(y, -expi(-exp(-y)), label=f'Gumbel')
plt.xlim(-2, 2)
plt.ylim(0, 1)
plt.legend()
plt.xlabel('$y$')
plt.ylabel(r'$\lambda F(\mathrm e^{-\lambda},(y+\ln\lambda)/\lambda)$')
plt.savefig('Gumbel.pdf', bbox_inches='tight', transparent=True, format='pdf')

def plot_F_k_2(k):
	x, F = x_and_F_k(k, upper_x = (2 + log(k))/k)
	y = x * k - log(k)
	plt.plot(y, F * k, label=r'$k=%i$' % k)

plt.figure(figsize=(2.8,2.1))
plot_F_k_2(5)
plot_F_k_2(20)
plot_F_k_2(35)
plot_F_k_2(50)
y = linspace(-2, 2, 201)
plt.plot(y, -expi(-exp(-y)), label=f'Gumbel')
plt.xlim(-2, 2)
plt.ylim(0, 1)
plt.legend()
plt.xlabel('$y$')
plt.ylabel(r'$kF_k((y+\ln k)/k)$')
plt.savefig('Gumbel_F_k.pdf', bbox_inches='tight', transparent=True, format='pdf')

File: plot/plot.rb
#!/usr/bin/env ruby

ENV['PYTHONPATH'] = Dir.glob(File.expand_path '../../lib/python3.*/site-packages', `readlink -f #{`which python`.strip}`.strip).first

require 'matplotlib/pyplot'

PLT = Matplotlib::Pyplot
PLT.rcParams.update({
	'text.usetex' => true,
	'font.size' => 10,
	'font.family' => 'lmodern',
	'text.latex.preamble' => '\usepackage{lmodern}'
})

def combo_pc m
	(1..m).each_with_object [[[1]]] do |n, dp|
		dp[n] = Array.new(n+1) { Array.new n+1, 0 }

		# dp[n][0] = (1-Y)**n
		0.upto(n-1) { dp[n][0][_1] = dp[n-1][0][_1] } # will be multiplied by 1-Y later

		1.upto n/2-1 do |k|
			# dp[n][k] = (1-Y) * (Y**k * (0..k).sum { |j| dp[n-k-1][j] } + (0..k-1).sum { |r| Y**r * dp[n-r-1][k] })
			0.upto(k) { |j| 0.upto(n-k-1) { dp[n][k][_1+k] += dp[n-k-1][j][_1] } }
			0.upto(k-1) { |r| 0.upto(n-r-1) { dp[n][k][_1+r] += dp[n-r-1][k][_1] } }
		end

		if n % 2 == 1
			k = n/2
			# dp[n][k] = (1-Y) * (Y**k + (0..k-1).sum { |r| Y**r * dp[n-r-1][k] })
			dp[n][k][k] = 1
			0.upto(k-1) { |r| 0.upto(n-r-1) { dp[n][k][_1+r] += dp[n-r-1][k][_1] } }
		end

		((n+1)/2).upto n-1 do |k|
			# dp[n][k] = (1-Y) * (Y**k + (0..n-k-1).sum { |r| Y**r * dp[n-r-1][k] })
			dp[n][k][k] = 1
			0.upto(n-k-1) { |r| 0.upto(n-r-1) { dp[n][k][_1+r] += dp[n-r-1][k][_1] } }
		end

		0.upto(n-1) { |k| n.downto(1) { dp[n][k][_1] -= dp[n][k][_1-1] } } # multiply by 1-Y
		
		# dp[n][n] = Y**n
		dp[n][n][n] = 1
	end.last
end

def horner coef, x
	coef.reverse.reduce(0) { |r, c| r * x + c }
end

def inv_coef a
	Array.new a.size do |j|
		binom = nil
		(j..N).sum do |k|
			a[k] * (binom = binom ? binom * k / (k-j) : 1)
		end * (-1)**j
	end
end

N = 50
COEF = combo_pc N
COEF_I = COEF.map { |a| inv_coef a }

def val k, y
	y < 0.5 ? horner(COEF[k], y) : horner(COEF_I[k], 1-y)
end

def plot y, label = '$p=%.3f$'%y
	xs = (0..N).to_a
	ys = xs.map { |k| val k, y }
	PLT.plot xs, ys, label:
end

def plot_finite
	PLT.figure figsize: [4.8, 3.6]
	plot 0.9
	plot 0.92
	plot 0.94
	plot 0.96
	plot 0.97
	plot 0.98
	plot 0.99
	plot 0.995
	plot 0.999
	PLT.ylim 0, 0.07
	PLT.legend
	PLT.xlabel '$k$'
	PLT.ylabel '$P_{%d,k}(p)$'%N
	PLT.savefig 'n50.pdf', transparent: true, format: :pdf, bbox_inches: :tight
end
plot_finite

include Math

class Integer
	def fac
		(1..self).reduce 1, :*
	end
end

def inf_first y, x
	-y**x*log(y)*(2+log(y**(x-1)))
end

def inf_rest y, x
	smax = x == 0.5 ? 3 : (1/x).ceil
	(2...smax).sum do |s|
		logysx1 = log y**(s*x-1)
		y**(s*x) * logysx1**(s-2) * s*(-1)**s/(s-2).fac * (
			1 + logysx1*2/(s-1) + logysx1**2/s/(s-1)
		)
	end * log(y)
end

def plot2 y, label = '$P=%.2f$'%y
	xs = Array.new(101) { |i| i/200.0 } + Array.new(101) { |i| 0.5+i/200.0 } + [1]
	ys1 = Array.new 81 do |i|
		x = i/200.0 + 0.1
		inf_first(y,x) + inf_rest(y,x)
	end
	ys2 = Array.new 101 do |i|
		x = i/200.0 + 0.5
		inf_first(y,x)
	end
	ys = [0]*20 + ys1 + ys2 + [5]
	PLT.plot xs, ys, label:
end

def plot_inf
	PLT.figure figsize: [4.8, 3.6]
	plot2 0.01
	plot2 0.03
	plot2 0.09
	plot2 0.3
	plot2 0.6
	plot2 0.9
	PLT.xlabel '$x$'
	PLT.ylabel '$f(P,x)$'
	PLT.ylim 0, 3
	PLT.legend
	PLT.savefig 'inf.pdf', transparent: true, format: :pdf, bbox_inches: :tight
end
plot_inf

def plot3 y
	y **= 1.0/N
	xs = (0..N).map { |k| k/N.to_f }
	ys = (0..N).map { |k| val(k, y)*N }
	PLT.plot xs, ys
end

def plot_compare
	PLT.figure figsize: [4.8, 3.6]
	plot3 0.01
	plot2 0.01
	plot3 0.2
	plot2 0.2
	plot3 0.5
	plot2 0.5
	plot3 0.9
	plot2 0.9
	PLT.xlabel '$x$'
	PLT.ylabel '$f(P,x)$ or $nP_{N,xn}(P^{1/n})$'
	PLT.ylim 0, 3
	PLT.legend
	PLT.title '$n=%d$'%N
	PLT.savefig 'compare.pdf', transparent: true, format: :pdf, bbox_inches: :tight
end
plot_compare
